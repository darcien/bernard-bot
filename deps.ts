// Sift is a small routing library that abstracts away details like starting a
// listener on a port, and provides a simple function (serve) that has an API
// to invoke a function for a specific path.
export {
  json,
  serve,
  Status,
  STATUS_TEXT,
  validateRequest,
} from "https://deno.land/x/sift@0.6.0/mod.ts";
// TweetNaCl is a cryptography library that we use to verify requests
// from Discord.
export { sign } from "https://cdn.skypack.dev/tweetnacl@v1.0.3?dts";

export * from "https://deno.land/x/discord_api_types@0.37.51/v10.ts";

export { markdownTable } from "https://esm.sh/markdown-table@3.0.3";

export { loadSync } from "https://deno.land/std@0.197.0/dotenv/mod.ts";

export { mapNotNullish } from "https://deno.land/std@0.197.0/collections/map_not_nullish.ts";
