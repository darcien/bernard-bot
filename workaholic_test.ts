import { assertEquals } from "https://deno.land/std@0.197.0/assert/mod.ts";
import { assertSnapshot } from "https://deno.land/std@0.197.0/testing/snapshot.ts";

import {
  formatWorkaholicAddCommand,
  isMessagePartialMatch,
  makeSummaryTable,
  parseMessageForSummary,
  WorkaholicType,
} from "./workaholic.ts";
import { APIMessage } from "./deps.ts";

Deno.test("isMessagePartialMatch", () => {
  assertEquals(
    isMessagePartialMatch("31 aug OT"),
    true,
  );
  assertEquals(
    isMessagePartialMatch("1 sep ot"),
    true,
  );
  assertEquals(
    isMessagePartialMatch("01 sep OT"),
    true,
  );
  assertEquals(
    isMessagePartialMatch("ot 2 jam"),
    true,
  );
  assertEquals(
    isMessagePartialMatch("OT 1hours"),
    true,
  );
});

Deno.test("isMessagePartialMatch", () => {
  assertEquals(
    isMessagePartialMatch("ot"),
    false,
  );
  assertEquals(
    isMessagePartialMatch("2 jam"),
    false,
  );
  assertEquals(
    isMessagePartialMatch("23 aug"),
    false,
  );
});

Deno.test("formatWorkaholicAddCommand", () => {
  const formatted = formatWorkaholicAddCommand({
    duration: 7,
    type: WorkaholicType.Overtime,
    userId: "123",
    what: "my lorem ipsum work",
    when: "2021-01-01",
  });
  assertEquals(
    formatted,
    "🐴 ⁃ <@123> ⁃ OT ⁃ 2021-01-01 ⁃ 7h ⁃ my lorem ipsum work",
  );
});

Deno.test("parseMessageForSummary", () => {
  const userInput = {
    duration: 7,
    type: WorkaholicType.Overtime,
    userId: "123",
    what: "my lorem ipsum work",
    when: "2021-01-01",
  };
  const formatted = formatWorkaholicAddCommand(userInput);
  const parsed = parseMessageForSummary({ content: formatted } as APIMessage);

  assertEquals(
    parsed,
    {
      ...userInput,
      duration: String(userInput.duration),
    },
  );

  const bad = parseMessageForSummary({ content: "🐴" } as APIMessage);
  assertEquals(
    bad,
    null,
  );
  assertEquals(
    parseMessageForSummary(
      { content: "🐴 ⁃ <@123> ⁃ OT ⁃ 2021-01-01 ⁃ 7h" } as APIMessage,
    ),
    null,
  );
  assertEquals(
    parseMessageForSummary(
      { content: "🐴 ⁃ <@123> ⁃ OT ⁃ 2021-01-01" } as APIMessage,
    ),
    null,
  );
  assertEquals(
    parseMessageForSummary(
      { content: "<@123> ⁃ OT ⁃ 2021-01-01" } as APIMessage,
    ),
    null,
  );
});

Deno.test("makeSummaryTable", async (t) => {
  const summary = makeSummaryTable({
    workaholicMessages: [
      {
        when: "2 aug",
        what: "do something cool",
        duration: "2h",
        type: WorkaholicType.Overtime,
        userId: "123",
      },
      {
        when: "3 aug",
        what: "fix something awesome",
        duration: "1h",
        type: WorkaholicType.Overtime,
        userId: "123",
      },
      {
        when: "7 aug",
        what: "prod incident something something",
        duration: "3h",
        type: WorkaholicType.Overtime,
        userId: "456",
      },
      {
        when: "17 aug",
        what: "its holiday and im working",
        duration: "8h",
        type: WorkaholicType.PriorityHours,
        userId: "456",
      },
      {
        when: "30 aug",
        what: "need to finish this thing for showcase",
        duration: "2h",
        type: WorkaholicType.Overtime,
        userId: "789",
      },
    ],
    nicknameByUserId: new Map([
      ["123", "Capybara"],
      ["456", "Yagi"],
      ["789", "Kaffu"],
    ]),
  });

  await assertSnapshot(t, summary);
});
