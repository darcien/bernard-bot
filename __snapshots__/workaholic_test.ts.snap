export const snapshot = {};

snapshot[`makeSummaryTable 1`] = `
"\`\`\`markdown
| Who      | When   | How Long | What                                   | Type |
| -------- | ------ | -------- | -------------------------------------- | ---- |
| Capybara | 2 aug  | 2h       | do something cool                      | OT   |
| Capybara | 3 aug  | 1h       | fix something awesome                  | OT   |
| Yagi     | 7 aug  | 3h       | prod incident something something      | OT   |
| Yagi     | 17 aug | 8h       | its holiday and im working             | PH   |
| Kaffu    | 30 aug | 2h       | need to finish this thing for showcase | OT   |
\`\`\`"
`;
